dynamiq/storages/vector/pgvector/pgvector.py

```diff
@@ -0,0 +1,551 @@
+
+        self.create_extension = create_extension
+        if self.create_extension:
+            self._conn.execute("CREATE EXTENSION IF NOT EXISTS vector")
+            self._conn.commit()
+
+        register_vector(self._conn)
+
+        self.client = client if client else self._conn
+
+        self.table_name = table_name
+        self.schema_name = schema_name
+        self.dimension = dimension
+        self.index_method = index_method
+        self.vector_function = vector_function
+
+        if create_if_not_exist:
+            with self._get_connection() as conn:
+                self._create_tables(conn)
+                if self.index_method in [PGVectorIndexMethod.IVFFLAT, PGVectorIndexMethod.HNSW]:
+                    self.index_name = index_name or f"{self.index_method}_index"
+                    self._create_index(conn)
+
+        logger.debug(f"PGVectorStore initialized with table_name: {self.table_name}")
+
+    @contextmanager
+    def _get_connection(self):
+        """Context manager for handling a single connection"""
+
+        import psycopg
+
+        if self._conn is None or self._conn.closed:
+            self._conn = psycopg.connect(self.connection_string)
+        try:
+            yield self._conn
+        except Exception as e:
+            self._conn.rollback()
+            raise e
+
+    def _execute_sql_query(
+        self, sql_query: Any, params: tuple | None = None, cursor: Cursor | None | None = None
+    ) -> Any:
+        """
+        Internal method to execute a SQL query.
+
+        Args:
+            sql_query (Any): The SQL query to execute.
+            params (tuple | None): The parameters to pass to the query. Defaults to None.
+            cursor (Optional[Cursor] | None): The cursor to use for the query. Defaults to None.
+
+        Raises:
+            VectorStoreException: If an error occurs while executing the query.
+
+        Returns:
+            Any: The result of the query.
+        """
+
+        params = params or ()
+
+        sql_query_str = sql_query.as_string(cursor) if not isinstance(sql_query, str) else sql_query
+
+        try:
+            result = cursor.execute(sql_query, params)
+        except Exception as e:
+            self._conn.rollback()
+            msg = f"Encountered an error while executing SQL query: {sql_query_str} with params: {params}. \nError: {e}"
+            raise VectorStoreException(msg)
+
+        return result
+
+    def _create_tables(self, conn) -> None:
+        """
+        Internal method to create the tables in the database (if they do not exist).
+
+        Args:
+            conn: The connection to the database.
+        """
+
+        query = SQL(
+            """
+            CREATE TABLE IF NOT EXISTS {schema_name}.{table_name} (
+                id VARCHAR(128) PRIMARY KEY,
+                content TEXT,
+                metadata JSONB,
+                embedding vector({dimension})
+            );
+            """
+        ).format(
```

Comment:
Please use native pcycopg params binding instead of create full str query with params by self. This helps to avoid SQL injections